include/pmacc/simulationControl/Checkpointing.hpp

```diff
@@ -0,0 +1,508 @@
+
+                /* create directory containing checkpoints  */
+                if(numCheckpoints == 0 && gc.getGlobalRank() == 0)
+                {
+                    pmacc::Filesystem::get().createDirectoryWithPermissions(checkpointDirectory);
+                }
+
+                Environment<DIM>::get().PluginConnector().checkpointPlugins(currentStep, checkpointDirectory);
+
+                /* important synchronize: only if no errors occured until this
+                 * point guarantees that a checkpoint is usable */
+                alpaka::wait(manager::Device<ComputeDevice>::get().current());
+
+                /* avoid deadlock between not finished PMacc tasks and MPI_Barrier */
+                eventSystem::getTransactionEvent().waitForFinished();
+
+                /* \todo in an ideal world with MPI-3, this would be an
+                 * MPI_Ibarrier call and this function would return a MPI_Request
+                 * that could be checked */
+                MPI_CHECK(MPI_Barrier(gc.getCommunicator().getMPIComm()));
+
+                if(gc.getGlobalRank() == 0)
+                {
+                    writeCheckpointStep(currentStep);
+                }
+                numCheckpoints++;
+            }
+        }
+
+        bool hasSoftRestartAttemptsLeft()
+        {
+            static uint32_t nthSoftRestart = 0;
+            if(nthSoftRestart <= softRestarts)
+            {
+                nthSoftRestart++;
+                return true;
+            }
+            return false;
+        }
+
+        void startTimeBasedCheckpointing()
+        {
+            // register concurrent thread to perform checkpointing periodically after a user defined time
+            if(checkpointPeriodMinutes != 0)
+                checkpointTimeThread = std::thread(
+                    [&, this]()
+                    {
+                        std::unique_lock<std::mutex> lk(this->concurrentThreadMutex);
+                        while(exitConcurrentThreads.wait_until(
+                                  lk,
+                                  std::chrono::system_clock::now() + std::chrono::minutes(checkpointPeriodMinutes))
+                              == std::cv_status::timeout)
+                        {
+                            signal::detail::setCreateCheckpoint(1);
+                        }
+                    });
+        }
+
+        void finishTimeBasedCheckpointing()
+        {
+            {
+                // notify all concurrent threads to exit
+                std::unique_lock<std::mutex> lk(this->concurrentThreadMutex);
+                exitConcurrentThreads.notify_all();
+            }
+            // wait for time triggered checkpoint thread
+            if(checkpointTimeThread.joinable())
+                checkpointTimeThread.join();
+        }
+
+        /* Returns whether a restart needs to be performed */
+        bool checkRestart(uint32_t& step)
```

Comment:
If you have that comment right above, why not?
```suggestion
        bool restartNeedsToBePerformed(uint32_t& step)
```

Reply:
`restartNeedsToBePerformed` also sounds like an instruction to do something

Reply:
I see... I was thinking about
```C++
if(checkpointing.restartNeedsToBePerformed)
```
which reads quite clearly but that might be just to my ears. This is not blocking. We can keep it as is.